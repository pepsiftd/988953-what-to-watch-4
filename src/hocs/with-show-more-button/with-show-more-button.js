import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

import {ShowMoreButton} from '@/components/show-more-button/show-more-button';

const withShowMoreButton = (Component) => {
  class WithShowMoreButton extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        cardsShowing: props.initialCardsCount,
        isButtonShowing: props.movies.length > props.initialCardsCount,
      };

      this.onShowMoreButtonClick = this.onShowMoreButtonClick.bind(this);
    }

    onShowMoreButtonClick() {
      this.setState((prevState) => ({
        cardsShowing: prevState.cardsShowing + this.props.addCardsOnShowMore,
        isButtonShowing: this.props.movies.length > (prevState.cardsShowing + this.props.addCardsOnShowMore),
      }));
    }

    componentDidUpdate(prevProps) {
      if (this.props.movies !== prevProps.movies) {
        this.setState({
          cardsShowing: this.props.initialCardsCount,
          isButtonShowing: this.props.movies.length > this.props.initialCardsCount,
        });
      } else {
        this.setState({
          isButtonShowing: this.props.movies.length > this.state.cardsShowing,
        });
      }
    }

    render() {
      return (
        <Component
          {...this.props}
          cardsShowing={this.state.cardsShowing}
          renderShowMore={() => {
            return this.state.isButtonShowing ? (
              <ShowMoreButton
                onClick={this.onShowMoreButtonClick}
              />
            ) : ``;
          }}
        />);
    }
  }

  WithShowMoreButton.propTypes = {
    movies: PropTypes.array.isRequired,
    initialCardsCount: PropTypes.number.isRequired,
    addCardsOnShowMore: PropTypes.number.isRequired,
  };

  return WithShowMoreButton;
};

export {withShowMoreButton};
