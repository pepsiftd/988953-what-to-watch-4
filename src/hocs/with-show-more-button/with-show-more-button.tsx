import React, {PureComponent} from 'react';
import {Subtract} from 'utility-types';

import {ShowMoreButton} from '@/components/show-more-button/show-more-button';

interface State {
  cardsShowing: number;
  isButtonShowing: boolean;
};

interface InjectingProps {
  cardsShowing: number;
  renderShowMore: () => React.ReactNode;
};

const withShowMoreButton = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithShowMoreButton extends PureComponent<T, State> {
    constructor(props) {
      super(props);

      this.state = {
        cardsShowing: props.initialCardsCount,
        isButtonShowing: props.movies.length > props.initialCardsCount,
      };

      this.onShowMoreButtonClick = this.onShowMoreButtonClick.bind(this);
    }

    onShowMoreButtonClick() {
      this.setState((prevState) => ({
        cardsShowing: prevState.cardsShowing + this.props.addCardsOnShowMore,
        isButtonShowing: this.props.movies.length > (prevState.cardsShowing + this.props.addCardsOnShowMore),
      }));
    }

    componentDidUpdate(prevProps) {
      if (this.props.movies !== prevProps.movies) {
        this.setState({
          cardsShowing: this.props.initialCardsCount,
          isButtonShowing: this.props.movies.length > this.props.initialCardsCount,
        });
      } else {
        this.setState({
          isButtonShowing: this.props.movies.length > this.state.cardsShowing,
        });
      }
    }

    render() {
      return (
        <Component
          {...this.props}
          cardsShowing={this.state.cardsShowing}
          renderShowMore={() => {
            return this.state.isButtonShowing ? (
              <ShowMoreButton
                onClick={this.onShowMoreButtonClick}
              />
            ) : ``;
          }}
        />);
    }
  }

  return WithShowMoreButton;
};

export {withShowMoreButton};
