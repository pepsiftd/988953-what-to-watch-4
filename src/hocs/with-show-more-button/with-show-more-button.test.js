import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';

import {withShowMoreButton} from './with-show-more-button';

const MockComponent = ({renderShowMore}) => {
  return (
    <div>
      {renderShowMore()}
    </div>
  );
};

MockComponent.propTypes = {
  renderShowMore: PropTypes.func.isRequired
};

it(`withShowMoreButton renders ShowMoreButton correctly`, () => {
  const MockComponentWrapped = withShowMoreButton(MockComponent);

  const tree = renderer.create(
      <MockComponentWrapped
        movies={[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
        initialCardsCount={4}
        addCardsOnShowMore={4}
      />
  );

  expect(tree).toMatchSnapshot();
});

it(`withShowMoreButton does not render ShowMoreButton when not needed`, () => {
  const MockComponentWrapped = withShowMoreButton(MockComponent);

  const tree = renderer.create(
      <MockComponentWrapped
        movies={[1, 2, 3]}
        initialCardsCount={4}
        addCardsOnShowMore={4}
      />
  );

  expect(tree).toMatchSnapshot();
});
