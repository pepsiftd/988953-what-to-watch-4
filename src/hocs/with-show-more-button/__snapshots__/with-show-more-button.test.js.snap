// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withShowMoreButton does not render ShowMoreButton when not needed 1`] = `
<div>
  
</div>
`;

exports[`withShowMoreButton renders ShowMoreButton correctly 1`] = `
<div>
  <div
    className="catalog__more"
  >
    <button
      className="catalog__button"
      onClick={[Function]}
      type="button"
    >
      Show more
    </button>
  </div>
</div>
`;
